File: August-LeetCoding-Challenge/31-Delete-Node-in-a-BST/Delete-Node-in-a-BST.cs
public class Solution {
    public TreeNode DeleteNode(TreeNode root, int key) {
        if(root == null) return null;
        if(key < root.val){
            root.left = DeleteNode(root.left, key);
        }else if(key > root.val){
            root.right = DeleteNode(root.right, key);
        }else{
            if(root.left == null)
            {
                return root.right;
            }
            else if(root.right == null){
                return root.left;
            }
            
            var minNode = FindMin(root.right);
            root.val = minNode.val;
            root.right = DeleteNode(root.right, root.val);
        }
        return root;
    }
    
    public TreeNode FindMin(TreeNode node){
        while(node.left != null){
            node = node.left;
        }
        return node;
    }
}
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

File: August-LeetCoding-Challenge/31-Delete-Node-in-a-BST/Delete-Node-in-a-BST.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} key
# @return {TreeNode}
def delete_node(root, key)
    if root == nil 
        return root
    elsif root.val > key
        root.left = delete_node(root.left, key)
        
    elsif root.val < key
        root.right = delete_node(root.right, key)
        
    else
        if root.left == nil && root.right == nil
            root = nil
            return nil
        elsif root.left == nil
            tmp = root.right
            root = nil
            return tmp
        elsif root.right == nil
            tmp = root.left 
            root = nil
            return tmp
        else
            tmp = min_value(root.right)
            root.val = tmp.val
            root.right = delete_node(root.right, tmp.val)
        end
    end
    return root
end

def min_value(node)
    current = node
    while current.left != nil
        current = current.left 
    end
    current
end